kafkaesk/retry.py

```diff
@@ -0,0 +1,220 @@
+
+        self._handlers[error] = handler
+
+        if self._initialized is True:
+            await self._handlers[error].initialize()
+
+        # Clear handler cache when handler is added
+        self._handler_cache = {}
+
+    async def remove_retry_handler(self, error: Type[Exception]) -> None:
+        if error in self._handlers:
+            handler = self._handlers[error]
+
+            del self._handlers[error]
+
+            # Clear handler cache when handler is removed
+            self._handler_cache = {}
+
+            await handler.finalize()
+
+    async def initialize(self) -> None:
+        await asyncio.gather(*[handler.initialize() for handler in self._handlers.values()])
+        self._initialized = True
+
+    async def finalize(self) -> None:
+        self._initialized = False
+        await asyncio.gather(*[handler.finalize() for handler in self._handlers.values()])
+
+    async def __call__(
+        self,
+        record: ConsumerRecord,
+        error: Exception,
+        retry_history: Optional[RetryHistory] = None,
+    ) -> None:
+        if self._initialized is not True:
+            raise RuntimeError("RetryPolicy is not initalized")
+
+        handler_key, handler = self._get_handler(error)
+
+        if handler is None or handler_key is None:
+            raise error from error
+
+        if retry_history is None:
+            retry_history = RetryHistory()
+
+        # Add information about this failure to the history
+        retry_history.failures.append(
+            FailureInfo(
+                error=error.__class__.__name__, handler_key=handler_key, timestamp=datetime.now()
+            )
+        )
+
+        await handler(self, handler_key, retry_history, record, error)
+
+    def _get_handler(self, error: Exception) -> Tuple[Optional[str], Optional["RetryHandler"]]:
+        error_type = error.__class__
+
+        handler_key, handler = self._handler_cache.get(error_type, (None, None))
+
+        if handler is None:
+            handler = self._handlers.get(error_type)
+            if handler is not None:
+                handler_key = error_type.__name__
+                self._handler_cache[error_type] = (handler_key, handler)
+
+        if handler is None:
+            for handler_error_type in self._handlers.keys():
+                if isinstance(error, handler_error_type):
+                    handler = self._handlers[handler_error_type]
+                    handler_key = handler_error_type.__name__
+                    self._handler_cache[error_type] = (handler_key, handler)
+                    break
+
+        return (handler_key, handler)
+
+
+class RetryHandler(ABC):
+    def __init__(self, stream_id: str) -> None:
+        self.stream_id = stream_id
+
+        self._initialized = False
+
+    async def initialize(self) -> None:
+        self._initialized = True
+
+    async def finalize(self) -> None:
+        self._initialized = False
```

Comment:
nit: resetting `_initialized` to false suggests we might re-initialize. would we support re-starting or re-using a `RetryHandler` after finalization?

Reply:
Right now no, I don't believe there is any method in place where a `RetryPolicy` would be re-initalized; a new `RetryPolicy` instance is created with every `SubscriptionConsumer`.  The main reason I clear the `_initilization` flag is so that an exception would be raised if something attempts to call `RetryPolicy.handle()` after the connections have been shut down.

Reply:
gotcha. it may be overkill, but splitting this logic into `_initialized` and `_finalized` flags might clarify the behavior.

or to rename to `_ready` ?

Reply:
I like `_ready` I think.

Technically there is nothing stopping a `RetryPolicy` from being initialized a second time after it has been shut down it's just that there is nothing implemented ATM that would cause that to happen.  It should work ATM.

Reply:
I've renamed `_initialized` to `_ready`.